### PR TITLE
Os executors

### PR DESCRIPTION
**Pull request by hkaiser**
This adds two types of executors: os_executors and attaching executors. It also fixes a whole set of problems related to thread binding, command-line handling, and config variable additions. This is all on a single PR as it is the result of our work related to a paper.
Further this introduces executor_information_traits to decouple functions needed to query executor properties.

### CODE REVIEW

File: src/runtime/threads/policies/affinity_data.cpp [hunk 1]
```diff
@@ -152,14 +152,20 @@ namespace hpx { namespace threads { namespace policies { namespace detail
         init_cached_pu_nums(num_system_pus, t);
     }
 
+    static mask_cref_type get_empty_machine_mask()
+    {
+        threads::mask_type m = threads::mask_type();
+        threads::resize(m, hardware_concurrency());
+        return m;
```
Comment (sithhell) [inline src/runtime/threads/policies/affinity_data.cpp]: This returns a reference to a temporary.

File: src/runtime/threads/policies/affinity_data.cpp [hunk 2]
```diff
@@ -152,14 +152,20 @@ namespace hpx { namespace threads { namespace policies { namespace detail
...
     mask_cref_type affinity_data::get_pu_mask(topology const& topology,
         std::size_t num_thread, bool numa_sensitive) const
     {
         // --hpx:bind=none disables all affinity
         if (threads::test(no_affinity_, num_thread))
         {
-            threads::mask_type m = threads::mask_type();
-            threads::resize(m, hardware_concurrency());
+            static mask_type m = get_empty_machine_mask();
             return m;
```
Comment (sithhell) [inline src/runtime/threads/policies/affinity_data.cpp]: Same here.

File: src/runtime/threads/policies/affinity_data.cpp (hunk 1 above)
Comment (hkaiser) [inline src/runtime/threads/policies/affinity_data.cpp]: Yes, my bad. Will fix.

File: src/runtime/threads/policies/affinity_data.cpp (hunk 2 above)
Comment (hkaiser) [inline src/runtime/threads/policies/affinity_data.cpp]: This is no temporary. Do I miss something?

Comment (sithhell) [inline src/runtime/threads/policies/affinity_data.cpp]: Didn't see the static ... my bad ...

**Pull request merged**